manager/scheduler/scheduler.go

```diff
@@ -286,17 +286,80 @@ func (s *Scheduler) deleteTask(ctx context.Context, t *api.Task) {
 	}
 }
 
+func reconcileTPRs(nodeRes *api.Resources, assigned, taken map[string]*api.ThirdPartyResource) {
+	for taskRes, taskVal := range assigned {
+		nodeVal := nodeRes.ThirdParty[taskRes]
+		if _, ok := taken[taskRes]; !ok {
+			taken[taskRes] = &api.ThirdPartyResource{}
+		}
+
+		takenVal := taken[taskRes]
+
+		reconcileTPR(nodeVal, taskVal, takenVal)
+	}
+}
+
+func reconcileTPR(nodeVal, taskVal, taken *api.ThirdPartyResource) {
+
+	switch tr := taskVal.Resource.(type) {
+	case *api.ThirdPartyResource_Integer:
+		nr, ok := nodeVal.Resource.(*api.ThirdPartyResource_Integer)
+		if !ok {
+			return
+		}
+
+		nr.Integer.Val -= tr.Integer.Val
+		taken.Resource = &api.ThirdPartyResource_Integer{
+			Integer: &api.Integer{Val: tr.Integer.Val}}
+
+	case *api.ThirdPartyResource_Set:
+		nr, ok := nodeVal.Resource.(*api.ThirdPartyResource_Set)
+		if !ok {
+			return
+		}
+
+		voidVal := &api.Set_Void{}
+		res := &api.ThirdPartyResource_Set{
+			Set: &api.Set{Val: make(map[string]*api.Set_Void)}}
+
+		for tval := range tr.Set.Val {
+			for nval := range nr.Set.Val {
+				if tval != nval {
+					continue
+				}
+				delete(nr.Set.Val, nval)
+			}
+
+			res.Set.Val[tval] = voidVal
+		}
+
+		taken.Resource = res
+	}
+}
+
 func (s *Scheduler) createOrUpdateNode(n *api.Node) {
 	nodeInfo, _ := s.nodeSet.nodeInfo(n.ID)
 	var resources api.Resources
 	if n.Description != nil && n.Description.Resources != nil {
+		fmt.Println("reconcile")
 		resources = *n.Description.Resources
```

Comment:
I think this should be changed to `resources = *n.Description.Resources.Copy()`, now that the `Resources` structure contains pointers (changes to the local copy of `resources` shouldn't have side effects on the actual `Node` object).